### PR TITLE
[FIX](type)fix matchExactType for complex type 

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx
fe matchExactType function should call type.matchTypes for its own logic, do not switch case to do special logic otherwise we may meet core in be like this.
 ```
F20231208 18:54:39.359673 680131 block.h:535] Check failed: _data_types[i]->is_nullable()  target type: Struct(l_info:Nullable(Array(Nullable(String)))) src type: Struct(col:Nullable(Array(Nullable(UInt8))))
*** Check failure stack trace: ***
    @     0x5584e952b926  google::LogMessage::SendToLog()
    @     0x5584e9527ef0  google::LogMessage::Flush()
    @     0x5584e952c169  google::LogMessageFatal::~LogMessageFatal()
    @     0x5584cf17201e  doris::vectorized::MutableBlock::merge_impl<>()
    @     0x5584ceac4b1d  doris::vectorized::MutableBlock::merge<>()
    @     0x5584d4dd7de3  doris::vectorized::VUnionNode::get_next_const()
    @     0x5584d4dd9a45  doris::vectorized::VUnionNode::get_next()
    @     0x5584bce469bd  std::__invoke_impl<>()
    @     0x5584bce466d0  std::__invoke<>()
    @     0x5584bce465c7  _ZNSt5_BindIFMN5doris8ExecNodeEFNS0_6StatusEPNS0_12RuntimeStateEPNS0_10vectorized5BlockEPbEPS1_St12_PlaceholderILi1EESC_ILi2EESC_ILi3EEEE6__callIS2_JOS4_OS7_OS8_EJLm0ELm1ELm2ELm3EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE
    @     0x5584bce46358  std::_Bind<>::operator()<>()
    @     0x5584bce46208  std::__invoke_impl<>()
    @     0x5584bce46178  _ZSt10__invoke_rIN5doris6StatusERSt5_BindIFMNS0_8ExecNodeEFS1_PNS0_12RuntimeStateEPNS0_10vectorized5BlockEPbEPS3_St12_PlaceholderILi1EESD_ILi2EESD_ILi3EEEEJS5_S8_S9_EENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESL_E4typeEOSM_DpOSN_
    @     0x5584bce45c18  std::_Function_handler<>::_M_invoke()
    @     0x5584bce6412f  std::function<>::operator()()
    @     0x5584bce56382  doris::ExecNode::get_next_after_projects()
    @     0x5584bce26218  doris::PlanFragmentExecutor::get_vectorized_internal()
    @     0x5584bce2431b  doris::PlanFragmentExecutor::open_vectorized_internal()
    @     0x5584bce22a96  doris::PlanFragmentExecutor::open()
    @     0x5584bce27c9d  doris::PlanFragmentExecutor::execute()
    @     0x5584bcbdb3f8  doris::FragmentMgr::_exec_actual()
    @     0x5584bcbf982f  doris::FragmentMgr::exec_plan_fragment()::$_0::operator()()
    @     0x5584bcbf9715  std::__invoke_impl<>()
    @     0x5584bcbf96b5  _ZSt10__invoke_rIvRZN5doris11FragmentMgr18exec_plan_fragmentERKNS0_23TExecPlanFragmentParamsERKSt8functionIFvPNS0_12RuntimeStateEPNS0_6StatusEEEE3$_0JEENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESH_E4typeEOSI_DpOSJ_
    @     0x5584bcbf942d  std::_Function_handler<>::_M_invoke()
    @     0x5584b9dfd883  std::function<>::operator()()
    @     0x5584bd6e3929  doris::FunctionRunnable::run()
    @     0x5584bd6cf8ce  doris::ThreadPool::dispatch_thread()
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

**Comment by BePPPower** [comment]
Please add some test cases

**Comment by amorynan** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

File: fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java [hunk 1]
```diff
@@ -88,13 +88,14 @@ public boolean matchesType(Type t) {
             return false;
         }
 
-        // Array(Null) is a virtual Array type, can match any Array(...) type
-        if (itemType.isNull() || ((ArrayType) t).getItemType().isNull()) {
-            return true;
+        if (!((ArrayType) t).getContainsNull() == getContainsNull()) {
```
Comment (morrySnow) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java:91]: nit: !=

File: fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java [hunk 2]
```diff
@@ -88,13 +88,14 @@ public boolean matchesType(Type t) {
...
+            return false;
         }
 
-        return itemType.matchesType(((ArrayType) t).itemType)
-                && (((ArrayType) t).containsNull || !containsNull);
+        if (!itemType.matchesType(((ArrayType) t).itemType)) {
+            return false;
+        }
+        return true;
```
Comment (morrySnow) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java:98]: nit: `return itemType.matchesType(((ArrayType) t).itemType)`

Comment (amorynan) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java:98]: make sense!

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionTypeDeducers.java [hunk 3]
```diff
@@ -23,26 +23,48 @@
...
-    public static Type deduce(String fnName, Type[] args) {
+    public static Type deduce(String fnName, int argIdx, Type[] args) {
         if (DEDUCERS.containsKey(fnName)) {
-            return DEDUCERS.get(fnName).deduce(args);
+            return DEDUCERS.get(fnName).deduce(argIdx, args);
         }
         return null;
     }
 
+    public static class ArrayElemFuncDeducer implements TypeDeducer {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionTypeDeducers.java:45]: i think u should add some comments for this function

**Comment by amorynan** [comment]
run pipelinex_p0

**Comment by amorynan** [comment]
run pipelinex_p0

**Review by morrySnow — DISMISSED** [review]

File: fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java [hunk 4]
```diff
@@ -88,13 +88,11 @@ public boolean matchesType(Type t) {
             return false;
         }
 
-        // Array(Null) is a virtual Array type, can match any Array(...) type
-        if (itemType.isNull() || ((ArrayType) t).getItemType().isNull()) {
-            return true;
+        if (((ArrayType) t).getContainsNull() != getContainsNull()) {
```
Comment (xiaokang) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java:91]: what's the sematics of getContainsNull() in ArrayType?

Comment (xiaokang) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java:91]: What's the difference to itemType.isNull() ?

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 5]
```diff
@@ -2252,22 +2253,6 @@ public static boolean matchExactType(Type type1, Type type2, boolean ignorePreci
                     return isSameDecimalTypeWithDifferentPrecision(((ScalarType) type2).decimalPrecision(),
                             ((ScalarType) type1).decimalPrecision());
                 }
-            } else if (type2.isArrayType()) {
```
Comment (xiaokang) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:2255]: It's moved to ArrayType.matchesType() but matchesType() is not invoked here.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/IsNullPredicate.java [hunk 6]
```diff
@@ -47,17 +47,14 @@ public static void initBuiltins(FunctionSet functionSet) {
...
             functionSet.addBuiltinBothScalaAndVectorized(ScalarFunction.createBuiltinOperator(IS_NOT_NULL,
                     null, Lists.newArrayList(t), Type.BOOLEAN, NullableMode.ALWAYS_NOT_NULLABLE));
+        }
+        // for array type
+        for (Type complexType : Lists.newArrayList(Type.ARRAY, Type.MAP, Type.GENERIC_STRUCT)) {
+            functionSet.addBuiltinBothScalaAndVectorized(ScalarFunction.createBuiltinOperator(IS_NULL, null,
+                    Lists.newArrayList(complexType), Type.BOOLEAN, NullableMode.ALWAYS_NOT_NULLABLE));
 
-            // for array type
-            for (Type complexType : Lists.newArrayList(Type.ARRAY, Type.MAP, Type.GENERIC_STRUCT)) {
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/IsNullPredicate.java:52]: Is it added mutiple times?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/StructLiteral.java [hunk 7]
```diff
@@ -95,6 +97,11 @@ public String getStringValueInFe() {
     @Override
     protected void toThrift(TExprNode msg) {
         msg.node_type = TExprNodeType.STRUCT_LITERAL;
+        ((StructType) type).getFields().forEach(v -> msg.setChildType(v.getType().getPrimitiveType().toThrift()));
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/StructLiteral.java:100]: How it works when the code is missing before?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 8]
```diff
@@ -386,16 +386,41 @@ public Function specializeTemplateFunction(Function templateFunction, Function r
...
+            for (; i < args.size(); i++) {
+                if (args.get(i).hasTemplateType()) {
                     hasTemplateType = true;
-                    args[i] = args[i].specializeTemplateType(requestFunction.getArgs()[i], specializedTypeMap, false,
-                            enableDecimal256);
+                    // if args[i] is template type, and requestFunction.getArgs()[i] NULL_TYPE, we need call function
+                    // deduce to get the specific type
+                    Type deduceType = requestFunction.getArgs()[i];
+                    if (requestFunction.getArgs()[i].isNull()
+                            || (requestFunction.getArgs()[i] instanceof ArrayType
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:401]: add () and indent to be more readable

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 9]
```diff
@@ -386,16 +386,41 @@ public Function specializeTemplateFunction(Function templateFunction, Function r
...
+                    } else {
+                        args.set(i, args.get(i).specializeTemplateType(requestFunction.getArgs()[i],
+                                specializedTypeMap, false, enableDecimal256));
+                    }
                 }
             }
+            // here need to support varArgs template according to request data
+            if (specializedFunction.hasVarArgs() && i < requestFunction.getNumArgs()) {
+                for (; i < requestFunction.getNumArgs(); i++) {
+                    if (requestFunction.getArgs()[i].isNull()) {
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:416]: add comment to explain

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 10]
```diff
@@ -448,7 +473,20 @@ public static boolean isCastMatchAllowed(Function desc, Function candicate) {
         final Type[] candicateArgTypes = candicate.getArgs();
         if (!(descArgTypes[0] instanceof ScalarType)
                 || !(candicateArgTypes[0] instanceof ScalarType)) {
-            if (candicateArgTypes[0] instanceof ArrayType || candicateArgTypes[0] instanceof MapType) {
+            if (candicateArgTypes[0] instanceof ArrayType) {
+                // match is exactly type. but for null type , with in array|map elem can not return true, because for
+                // be will make null_type to uint8
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:478]: BE

File: fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java (hunk 4 above)
Comment (amorynan) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/ArrayType.java:91]: getContainsNull()  means nullable(Array)
getItemType().isNull() means current item is null_type

**Comment by amorynan** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Comment by amorynan** [comment]
run external

**Review by morrySnow — APPROVED** [review]

**Pull request merged**